cmd/acr/purge.go

```diff
@@ -463,32 +467,46 @@ func getManifestsToDelete(ctx context.Context, acrClient api.AcrCLIClientInterfa
 		return nil, err
 	}
 	// This will act as a set if a key is present then it should not be deleted because it is referenced by a multiarch manifest
-	// that will not be deleted
-	doNotDelete := map[string]bool{}
-	candidatesToDelete := []acr.ManifestAttributesBase{}
-	// Iterate over all manifests to discover multiarchitecture manifests
+	// or the manifest has subjects attached
+	doNotDelete := make(map[string]struct{})
+	var candidatesToDelete []acr.ManifestAttributesBase
 	for resultManifests != nil && resultManifests.ManifestsAttributes != nil {
 		manifests := *resultManifests.ManifestsAttributes
 		for _, manifest := range manifests {
-			if (*manifest.MediaType == manifestListContentType || *manifest.MediaType == manifestIndexContentType) && manifest.Tags != nil {
-				// If a manifest list is found and it has tags then all the dependent digests are
-				// marked to not be deleted.
-				var manifestListBytes []byte
-				manifestListBytes, err = acrClient.GetManifest(ctx, repoName, *manifest.Digest)
+			if _, ok := mediaTypes[*manifest.MediaType]; ok {
+				manifestBytes, err := acrClient.GetManifest(ctx, repoName, *manifest.Digest)
 				if err != nil {
 					return nil, err
 				}
-				var manifestList multiArchManifest
-				err = json.Unmarshal(manifestListBytes, &manifestList)
+				var customManifest customManifest
+				err = json.Unmarshal(manifestBytes, &customManifest)
 				if err != nil {
 					return nil, err
 				}
-				for _, dependentDigest := range manifestList.Manifests {
-					doNotDelete[dependentDigest.Digest] = true
+				if customManifest.Subject != nil {
+					// if manifest has subject, manifest is marked to not be deleted
+					doNotDelete[*manifest.Digest] = struct{}{}
+				} else if manifest.Tags == nil {
+					// If the manifest has no subject or tag, it is a candidate for deletion
+					candidatesToDelete = append(candidatesToDelete, manifest)
+				}
+				if len(customManifest.Manifests) > 0 {
```

Comment:
I would suggest verifying if this is a list directly instead of using the manifests field to check type. Instead, we can use a MediaType check which provides better readability.

Reply:
done. factor out the type checking process to a new method called filterManifests, which uses switch statements for MediaTypes.